lib/assert.js

```diff
@@ -71,11 +75,86 @@ assert.fail = fail;
 //                             expected: expected });
 assert.AssertionError = errors.AssertionError;
 
+function getBuffer(fd, assertLine) {
+  var lines = 0;
+  // Prevent blocking the event loop by limiting the maximum amount of
+  // data that may be read.
+  var maxReads = 64; // bytesPerRead * maxReads = 512 kb
+  var bytesRead = 0;
+  var startBuffer = 0; // Start reading from that char on
+  const bytesPerRead = 8192;
+  const buffers = [];
+  do {
+    const buffer = Buffer.allocUnsafe(bytesPerRead);
+    bytesRead = readSync(fd, buffer, 0, bytesPerRead);
+    for (var i = 0; i < bytesRead; i++) {
+      if (buffer[i] === 10) {
+        lines++;
+        if (lines === assertLine) {
+          startBuffer = i + 1;
+        // Read up to 15 more lines to make sure all code gets matched
+        } else if (lines === assertLine + 16) {
+          buffers.push(buffer.slice(startBuffer, i));
+          return buffers;
+        }
+      }
+    }
+    if (lines >= assertLine) {
+      buffers.push(buffer.slice(startBuffer, bytesRead));
+      // Reset the startBuffer in case we need more than one chunk
+      startBuffer = 0;
+    }
+  } while (--maxReads !== 0 && bytesRead !== 0);
+  return buffers;
+}
 
 // Pure assertion tests whether a value is truthy, as determined
 // by !!value.
 function ok(value, message) {
-  if (!value) innerFail(value, true, message, '==', ok);
+  if (!value) {
+    if (message == null) {
+      // Use the call as error message if possible.
+      // This does not work with e.g. the repl.
+      const tmp = Error.prepareStackTrace;
+      Error.prepareStackTrace = (_, stack) => stack;
+      const err = new Error();
+      Error.captureStackTrace(err, ok);
+      const call = err.stack[0];
+      Error.prepareStackTrace = tmp;
+      // Do not try to get the error if there is no code that we can read.
+      var fd;
+      try {
+        fd = openSync(call.getFileName(), 'r', 0o666);
```

Comment:
Open question: is there any chance that the string containing the file contents is actually already in memory somewhere, due to the module having been loaded?

Reply:
Not that I am aware of. The code is imported once and discarded right afterwards again when loading a module. That is also logical as the code might be quite big and could potentially block a big memory chunk. It would of course be nice to prevent reading the files but that is not possible currently (as far as I see it).

What I could do is adding a cache for errors that occurred before. That way each assert that triggers will only result in a single read instead of one per call.